In [1]:
from pandas import pandas as pd
movies = pd.read_csv("movies.csv")
ratings = pd.read_csv("ratings.csv")

In [2]:
movies.describe(include='all')

movieId        title genres
count     9742.000000         9742   9742
unique            NaN         9737    951
top               NaN  Emma (1996)  Drama
freq              NaN            2   1053
mean     42200.353623          NaN    NaN
std      52160.494854          NaN    NaN
min          1.000000          NaN    NaN
25%       3248.250000          NaN    NaN
50%       7300.000000          NaN    NaN
75%      76232.000000          NaN    NaN
max     193609.000000          NaN    NaN

In [3]:
ratings.describe(include='all')

userId        movieId         rating     timestamp
count  100836.000000  100836.000000  100836.000000  1.008360e+05
mean      326.127564   19435.295718       3.501557  1.205946e+09
std       182.618491   35530.987199       1.042529  2.162610e+08
min         1.000000       1.000000       0.500000  8.281246e+08
25%       177.000000    1199.000000       3.000000  1.019124e+09
50%       325.000000    2991.000000       3.500000  1.186087e+09
75%       477.000000    8122.000000       4.000000  1.435994e+09
max       610.000000  193609.000000       5.000000  1.537799e+09

In [4]:
df = pd.merge(movies, ratings, on='movieId')
df.describe(include='all')

movieId                title  genres         userId  \
count   100836.000000               100836  100836  100836.000000   
unique            NaN                 9719     951            NaN   
top               NaN  Forrest Gump (1994)  Comedy            NaN   
freq              NaN                  329    7196            NaN   
mean     19435.295718                  NaN     NaN     326.127564   
std      35530.987199                  NaN     NaN     182.618491   
min          1.000000                  NaN     NaN       1.000000   
25%       1199.000000                  NaN     NaN     177.000000   
50%       2991.000000                  NaN     NaN     325.000000   
75%       8122.000000                  NaN     NaN     477.000000   
max     193609.000000                  NaN     NaN     610.000000   

               rating     timestamp  
count   100836.000000  1.008360e+05  
unique            NaN           NaN  
top               NaN           NaN  
freq              NaN           NaN  
mean         3.501557  1.205946e+09  
std          1.042529  2.162610e+08  
min          0.500000  8.281246e+08  
25%          3.000000  1.019124e+09  
50%          3.500000  1.186087e+09  
75%          4.000000  1.435994e+09  
max          5.000000  1.537799e+09

Tanto df como ratings tienen 100836 rows. con eso compruebo que no perdi datos

filtro pelis mayores a 4

In [5]:
df = df[df['rating']>=4.0]
df.head()

movieId             title                                       genres  \
0        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
1        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
2        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
4        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
6        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   

   userId  rating   timestamp  
0       1     4.0   964982703  
1       5     4.0   847434962  
2       7     4.5  1106635946  
4      17     4.5  1305696483  
6      19     4.0   965705637

In [6]:
apriori_df = df.groupby('userId')['title'].apply(list)
apriori_df

userId
1      [Toy Story (1995), Grumpier Old Men (1995), He...
2      [Tommy Boy (1995), Good Will Hunting (1997), G...
3      [Escape from L.A. (1996), Conan the Barbarian ...
4      [Nobody Loves Me (Keiner liebt mich) (1994), F...
5      [Toy Story (1995), Get Shorty (1995), Babe (19...
                             ...                        
606    [Sense and Sensibility (1995), Four Rooms (199...
607    [Toy Story (1995), Dead Man Walking (1995), Wh...
608    [GoldenEye (1995), Casino (1995), Seven (a.k.a...
609    [GoldenEye (1995), Interview with the Vampire:...
610    [Toy Story (1995), Heat (1995), Casino (1995),...
Name: title, Length: 609, dtype: object

In [9]:
from mlxtend.frequent_patterns import apriori, association_rules

In [10]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder

te = TransactionEncoder()
te_ary = te.fit(apriori_df).transform(apriori_df)
df2 = pd.DataFrame(te_ary, columns=te.columns_)
df2

'71 (2014)  'Hellboy': The Seeds of Creation (2004)  'Salem's Lot (2004)  \
0         False                                    False                False   
1         False                                    False                False   
2         False                                    False                False   
3         False                                    False                False   
4         False                                    False                False   
..          ...                                      ...                  ...   
604       False                                    False                False   
605       False                                    False                False   
606       False                                    False                False   
607       False                                    False                False   
608        True                                    False                False   

     'Til There Was You (1997)  'burbs, The (1989)  \
0                        False               False   
1                        False               False   
2                        False               False   
3                        False               False   
4                        False               False   
..                         ...                 ...   
604                      False               False   
605                      False               False   
606                      False               False   
607                      False               False   
608                      False               False   

     (500) Days of Summer (2009)  *batteries not included (1987)  \
0                          False                           False   
1                          False                           False   
2                          False                           False   
3                          False                           False   
4                          False                           False   
..                           ...                             ...   
604                        False                           False   
605                        False                           False   
606                        False                           False   
607                        False                           False   
608                        False                           False   

     ...And Justice for All (1979)  \
0                            False   
1                            False   
2                            False   
3                            False   
4                            False   
..                             ...   
604                          False   
605                          False   
606                          False   
607                          False   
608                          False   

     00 Schneider - Jagd auf Nihil Baxter (1994)  1-900 (06) (1994)  ...  \
0                                          False              False  ...   
1                                          False              False  ...   
2                                          False              False  ...   
3                                          False              False  ...   
4                                          False              False  ...   
..                                           ...                ...  ...   
604                                        False              False  ...   
605                                        False              False  ...   
606                                        False              False  ...   
607                                        False              False  ...   
608                                        False              False  ...   

     Zombieland (2009)  Zookeeper (2011)  Zoolander (2001)  Zootopia (2016)  \
0                False             False             False            False   
1                False             Fal

In [11]:
f=apriori(df2, min_support=0.15, max_len=2, use_colnames=True)
#Esto es una variante a lo que vimos la clase anterior
rules=association_rules(f,metric='lift',min_threshold=2)

In [12]:
rules

antecedents  \
0                             (Godfather, The (1972))   
1                    (Godfather: Part II, The (1974))   
2   (Lord of the Rings: The Fellowship of the Ring...   
3   (Lord of the Rings: The Return of the King, Th...   
4     (Lord of the Rings: The Two Towers, The (2002))   
5   (Lord of the Rings: The Fellowship of the Ring...   
6     (Lord of the Rings: The Two Towers, The (2002))   
7   (Lord of the Rings: The Return of the King, Th...   
8                                (Matrix, The (1999))   
9                        (Saving Private Ryan (1998))   
10                               (Matrix, The (1999))   
11                          (Sixth Sense, The (1999))   
12        (Star Wars: Episode IV - A New Hope (1977))   
13  (Raiders of the Lost Ark (Indiana Jones and th...   
14  (Star Wars: Episode V - The Empire Strikes Bac...   
15  (Raiders of the Lost Ark (Indiana Jones and th...   
16  (Star Wars: Episode V - The Empire Strikes Bac...   
17        (Star Wars: Episode IV - A New Hope (1977))   
18  (Star Wars: Episode VI - Return of the Jedi (1...   
19        (Star Wars: Episode IV - A New Hope (1977))   
20  (Star Wars: Episode VI - Return of the Jedi (1...   
21  (Star Wars: Episode V - The Empire Strikes Bac...   

                                          consequents  antecedent support  \
0                    (Godfather: Part II, The (1974))            0.259442   
1                             (Godfather, The (1972))            0.175698   
2   (Lord of the Rings: The Return of the King, Th...            0.239737   
3   (Lord of the Rings: The Fellowship of the Ring...            0.229885   
4   (Lord of the Rings: The Fellowship of the Ring...            0.215107   
5     (Lord of the Rings: The Two Towers, The (2002))            0.239737   
6   (Lord of the Rings: The Return of the King, Th...            0.215107   
7     (Lord of the Rings: The Two Towers, The (2002))            0.229885   
8                        (Saving Private Ryan (1998))            0.364532   
9                                (Matrix, The (1999))            0.231527   
10                          (Sixth Sense, The (1999))            0.364532   
11                               (Matrix, The (1999))            0.205255   
12  (Raiders of the Lost Ark (Indiana Jones and th...            0.330049   
13        (Star Wars: Episode IV - A New Hope (1977))            0.267652   
14  (Raiders of the Lost Ark (Indiana Jones and th...            0.275862   
15  (Star Wars: Episode V - The Empire Strikes Bac...            0.267652   
16        (Star Wars: Episode IV - A New Hope (1977))            0.275862   
17  (Star Wars: Episode V - The Empire Strikes Bac...            0.330049   
18        (Star Wars: Episode IV - A New Hope (1977))            0.246305   
19  (Star Wars: Episode VI - Return of the Jedi (1...            0.330049   
20  (Star Wars: Episode V - The Empire Strikes Bac...            0.246305   
21  (Star Wars: Episode VI - Return of the Jedi (1...            0.275862   

    consequent support   support  confidence      lift  leverage  conviction  
0             0.175698  0.167488    0.645570  3.674317  0.121904    2.325710  
1             0.259442  0.167488    0.953271  3.674317  0.121904   15.847947  
2             0.229885  0.198686    0.828767  3.605137  0.143574    4.497471  
3             0.239737  0.198686    0.864286  3.605137  0.143574    5.601936  
4             0.239737  0.183908    0.854962  3.566245  0.132339    5.241811  
5             0.215107  0.183908    0.767123  3.566245  0.132339    3.370424  
6             0.229885  0.192118    0.893130  3.885115  0.142668    7.206076  
7             0.215107  0.192118    0.835714  3.885115  0.142668    4.777611  
8             0.231527  0.177340    0.486486  2.101208  0.092941    1.496500  
9             0.364532  0.177340    0.765957  2.101208  0.092941    2.715181  
10            0.205255  0.151067    0.414414  2.019027  0.076245    1.357181  
11        

In [13]:
from IPython.display import *
import ipywidgets as widgets
from ipywidgets import *

In [14]:
raw_input = None #@param {type:"raw"}

In [15]:
print(raw_input)

None
